de.variantsync.core/src/main/java/de/variantsync/core/ast/AST.java

```diff
@@ -0,0 +1,183 @@
+package de.variantsync.core.ast;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.UUID;
+
+import de.variantsync.core.interfaces.Grammar;
+
+/**
+ * This class represents the Abstract Syntax Tree data structure. This also tests the LineGrammar indirectly.
+ *
+ * @param <G> a generic which needs to extend the Grammar class, defining the type of the AST
+ * @param <V> a generic which defines the value of the actual AST
+ * @author eric
+ */
+public class AST<G extends Grammar, V> {
+
+	private UUID id;
+	private G type;
+	private V value;
+	private List<AST<G, V>> subtrees;
+
+	// all attributes which should not be visible to the GSON parser need to be at least transient
+	public static transient final String INDENT_STRING = "    ";
+	public static transient final String NEXT_SEPARATOR = "\u2502 ";
+	public static transient final String NEXT_ACT_SEPARATOR = "\u251C\u2500 ";
+	public static transient final String LAST_SEPARATOR = "\u2514\u2500 ";
+
+	public AST(UUID id, G type, V value) {
+		this.id = id;
+		this.type = type;
+		this.value = value;
+		this.subtrees = new ArrayList<>();
+	}
+
+	public AST(G type, V value) {
+		this(UUID.randomUUID(), type, value);
+	}
+
+	/**
+	 * Empty AST is forbidden at the moment.
+	 */
+	private AST() {
+
+	}
+
+	public UUID getId() {
+		return id;
+	}
+
+	public V getValue() {
+		return value;
+	}
+
+	public G getType() {
+		return type;
+	}
+
+	/**
+	 * @return the subtrees as an unmodifiable List
+	 */
+	public List<AST<G, V>> getSubtrees() {
+		return Collections.unmodifiableList(subtrees);
+	}
+
+	public int getMaxDepth() {
+		if (subtrees.size() == 0) {
+			return 0;
+		}
+
+		int maxDepth = 0;
+		for (final AST<G, V> node : subtrees) {
+			maxDepth = Math.max(node.getMaxDepth(), maxDepth);
+		}
+		return ++maxDepth;
+	}
+
+	/**
+	 * This method calls the addChild method for each element of the given list. ASTs which would be rejected by the addChild method are skipped.
+	 *
+	 * @param toAdd List of AST which should be added as subtrees
+	 * @return true if all items where successfully added
+	 */
+	public boolean addChildren(List<AST<G, V>> toAdd) {
+		boolean out = true;
+		if (toAdd != null) {
+			for (final AST<G, V> elem : toAdd) {
+				if (!addChild(elem)) {
+					out = false;
+				}
+			}
+		}
+		return out;
```

Comment:
Returns `true` if `toAdd == null`, but better should return `false` meaning the argument could not be added.